src/sgame/sg_cmds.cpp

```diff
@@ -2252,6 +2252,56 @@ static bool FindRoomForClassChangeVertically(
 	return !tr.startsolid && tr.fraction == 1.0f;
 }
 
+static bool FindRoomForClassChangeLaterally(
+		const gentity_t *ent,
+		const vec3_t fromMins, const vec3_t fromMaxs,
+		const vec3_t toMins, const vec3_t toMaxs,
+		vec3_t newOrigin)
+{
+	float max_x = fabs(fabs(toMaxs[0]-toMins[0])
+	                 - fabs(fromMaxs[0]-fromMins[0]));
+	float max_y = fabs(fabs(toMaxs[1]-toMins[1])
+	                 - fabs(fromMaxs[1]-fromMins[1]));
+	// this is an upper bound, as small as possible to avoid
+	// loosing valid results. ideally, it would be the correct minimal
+	// length for each 4 directions, but this is close enough.
+	float distance = std::max(max_x, max_y);
+
+	vec3_t origin;
+	origin[0] = ent->client->ps.origin[0];
+	origin[1] = ent->client->ps.origin[1];
+	origin[2] = ent->client->ps.origin[2] + fromMins[2] - toMins[2] + 1.0f;
+
+	float best_distance = 1e37f;
+	bool found = false;
+	vec3_t deltas[] = { { 1, 0 }, { 0, 1 }, { -1, 0 }, { 0, -1 },
```

Comment:
This is probably going to give warnings for having less than 3 elements.

Reply:
fixed